### PR TITLE
[PIR] change IR clone to const and support clone operation successors

### PR DESCRIPTION
**Pull request by huangjiyi**
### PR types

Others

### PR changes

Others

### Description

- #60590 的后续 PR
- 将 IR 各模块类的 clone 接口改为 const 接口
- 添加了 Operation successors clone 的实现

### CODE REVIEW

File: paddle/pir/core/ir_mapping.h [hunk 1]
```diff
@@ -21,25 +21,46 @@ class Block;
...
+
 template <typename T, typename... OthersT>
 struct ExactlyOneIrType {
   using type = void;
 };
 template <typename T, typename FirstT, typename... OthersT>
 struct ExactlyOneIrType<T, FirstT, OthersT...> {
   using type =
-      std::conditional_t<std::is_convertible<T, FirstT>::value,
+      std::conditional_t<std::is_convertible<FirstT, ConstIrType<T>>::value,
```
Comment (huangjiyi) [inline paddle/pir/core/ir_mapping.h:46]: 这里是因为 `std::conditional_t<Value, OpResult>::value == false`

File: paddle/pir/core/ir_mapping.h [hunk 2]
```diff
@@ -21,25 +21,46 @@ class Block;
...
                          FirstT,
                          typename ExactlyOneIrType<T, OthersT...>::type>;
 };
```
Comment (huangjiyi) [inline paddle/pir/core/ir_mapping.h:49]: 这里是因为 `std::conditional_t<Value, OpResult>::value == false`

File: paddle/pir/core/ir_mapping.h [hunk 3]
```diff
@@ -62,15 +84,16 @@ class IrMapping {
...
   template <typename T, typename S>
   void Add(T from, S to) {
     if (!from) return;
     GetMutableMap<IrType<T>>()[from] = to;
   }
 
   template <typename T>
-  T Lookup(T from) const {
-    if (!from) return static_cast<T>(nullptr);
+  IrType<T> Lookup(T from) const {
```
Comment (winter-wang) [inline paddle/pir/core/ir_mapping.h:95]: T Lookup(const T& from) const {
}

Comment (winter-wang) [inline paddle/pir/core/ir_mapping.h:83]:
```suggestion
  IrType<T> Lookup(const T & from) const {
```

**Review by winter-wang — APPROVED** [review]

**Pull request merged**